### PR TITLE
Dev cito to

### PR DESCRIPTION
**Pull request by FFkmmr**

### CODE REVIEW

File: my_set/main/static/csv/data.csv
Comment (Quentas) [inline my_set/main/static/csv/data.csv:1]: Such files should not be saved to repository

File: my_set/main/management/commands/import_projects.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+                    description=row['description'].strip(), 
+                )
+
+                for tech_name in technologies:
+                    technology = Technology.objects.get_or_create(name=tech_name.strip())
+                    project.technologies.add(technology)
+
+                for industry_name in industries:
+                    industry = Industry.objects.get_or_create(name=industry_name.strip())
+                    project.industries.add(industry)
```
Comment (Quentas) [inline my_set/main/management/commands/import_projects.py]: Parsing and saving functoinalities should be moved to the other file, like `utils.py`, `helpers.py` or `service.py`.

File: my_set/main/models.py [hunk 2]
```diff
@@ -1,3 +1,27 @@
...
+    def num_projects(self):
+        return self.project_set.count()
+
+class Project(models.Model):
+    title = models.CharField(max_length=50)
+    url = models.CharField(max_length=100)
+    technologies = models.ManyToManyField(Technology)
+    description = models.TextField()
+    industries = models.ManyToManyField(Industry)
+
```
Comment (Quentas) [inline my_set/main/models.py:37]: Project should be assigned to specific user

File: my_set/main/static/main/img/image.png
Comment (Quentas) [inline my_set/main/static/main/img/image.png:1]: Such files should not be included to repository

File: my_set/main/urls.py [hunk 3]
```diff
@@ -3,7 +3,10 @@
...
     path('', views.index, name='home'),
-    path('register/', views.registerPage, name='register'),
-    path('login/', views.loginPage, name='login'),
-    path('logout/', views.loginPage, name='logout'),
+    path('project_list/', views.project_filter_view, name='project_list'),
+    path('register/', views.register_page, name='register'),
+    path('login/', views.login_page, name='login'),
+    path('logout/', views.logout_user, name='logout'),
+    path('import_csv/', views.import_csv_view, name='import_csv'),
+    path('add-project/', views.add_project, name="add_project"),
```
Comment (Quentas) [inline my_set/main/urls.py]: Endpoint urls should inlcude dash (`-`) rather then underscore (`_`) as a separator between words.

File: my_set/main/views.py [hunk 4]
```diff
@@ -38,17 +39,108 @@ def loginPage(request):
...
+        **filters
+    )
+
+    context = {
+        'projects': projects,
+    }
+
+    html = render_to_string('main/projects.html', context)
+    return JsonResponse({'html': html})
+@login_required(login_url='login')
```
Comment (Quentas) [inline my_set/main/views.py]: Add space

File: my_set/main/views.py [hunk 5]
```diff
@@ -38,17 +39,108 @@ def loginPage(request):
...
+                            technology, created = Technology.objects.get_or_create(name=tech_name.strip())
+                            technologies.append(technology)
+                        
+                        industries = []
+                        for industry_name in industry_names:
+                            industry, created = Industry.objects.get_or_create(name=industry_name.strip())
+                            industries.append(industry)
+                        
+                        project.technologies.set(technologies)
+                        project.industries.set(industries)
```
Comment (Quentas) [inline my_set/main/views.py]: Move this functonality to separate file

**Review by Quentas — CHANGES_REQUESTED** [review]

File: my_set/main/forms.py [hunk 6]
```diff
@@ -1,7 +1,49 @@
...
+            'description': forms.Textarea(attrs={'placeholder': 'Enter project description...', 'rows': 4}),
+            'technologies': forms.CheckboxSelectMultiple(),
+            'industries': forms.CheckboxSelectMultiple(),
+            'new_technologies': forms.TextInput(attrs={'placeholder': 'Enter new technologies, separated by commas...',
+            'class': 'wide-input'}),
+            'new_idustries': forms.TextInput(attrs={'placeholder': 'Enter new industries, separated by commas...',
+            'class': 'wide-input'}),
+            'sets': forms.CheckboxSelectMultiple(),
+        }
+class CreateProjectSet(ModelForm):
```
Comment (Quentas) [inline my_set/main/forms.py:42]: add two empty lines above

File: my_set/main/forms.py [hunk 7]
```diff
@@ -1,7 +1,49 @@
...
+        'email': 'Email..',
+        'password1': 'Enter password...',
+        'password2': 'Re-enter Password...'
+    }
+
+    def __init__(self, *args, **kwargs):
+        super(CreateUserForm, self).__init__(*args, **kwargs)
+        for field_name, placeholder in self.placeholders.items():
+            self.fields[field_name].widget.attrs.update({'placeholder': placeholder})
+class CreateProjectForm(ModelForm):
```
Comment (Quentas) [inline my_set/main/forms.py:24]: add two empty lines above

File: my_set/main/models.py [hunk 8]
```diff
@@ -1,3 +1,48 @@
...
+class Industry(models.Model):
+    name = models.CharField(max_length=50)
+    def __str__(self):
+        return self.name
+    @property
+    def num_projects(self):
+        return self.project_set.count()
+    
+class Sets(models.Model):
```
Comment (Quentas) [inline my_set/main/models.py:20]: name of a model should be in single form

File: my_set/main/models.py [hunk 9]
```diff
@@ -1,3 +1,48 @@
...
+        return self.name
+    @property
+    def num_projects(self):
+        return self.project_set.count()
+
+class Industry(models.Model):
+    name = models.CharField(max_length=50)
+    def __str__(self):
+        return self.name
+    @property
```
Comment (Quentas) [inline my_set/main/models.py:16]: add empty line above, in all such places

File: my_set/main/models.py (hunk 8 above)
Comment (Quentas) [inline my_set/main/models.py:20]: what is the purpose of this model, if you have Sets?

File: my_set/main/management/commands/service.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+                    description=row['description'].strip(), 
+                )
+
+                for tech_name in technologies:
+                    technology = Technology.objects.get_or_create(name=tech_name.strip())
+                    project.technologies.add(technology)
+    
+                for industry_name in industries:
+                    industry = Industry.objects.get_or_create(name=industry_name.strip())
+                    project.industries.add(industry)
```
Comment (Quentas) [inline my_set/main/management/commands/service.py:28]: move this to other file, not management/commands/service.py. Better to main/service.py

File: my_set/main/views.py [hunk 11]
```diff
@@ -38,17 +43,234 @@ def loginPage(request):
...
+                        technologies = []
+                        for tech_name in tech_names:
+                            technology, created = Technology.objects.get_or_create(name=tech_name.strip())
+                            technologies.append(technology)
+                        industries = []
+                        for industry_name in industry_names:
+                            industry, created = Industry.objects.get_or_create(name=industry_name.strip())
+                            industries.append(industry)
+                        project.technologies.set(technologies)
+                        project.industries.set(industries)
```
Comment (Quentas) [inline my_set/main/views.py:134]: move this to main/service.py

File: my_set/main/views.py [hunk 12]
```diff
@@ -38,17 +43,234 @@ def loginPage(request):
...
+        email_send(email, link)
+        return redirect('home')
+
+    return render(request, 'template.html')
+
+def email_send(email, link):
+  	return requests.post(
+  		"https://api.mailgun.net/v3/sandbox4694cce68b6743b6ab979c5af9994e42.mailgun.org/messages",
+  		auth=("api", "eaea9772c8e26ff0a5966571ed60336f-623e10c8-6896f4ca"),
+  		data={"from": "mitsuhanocreations@gmail.com",
```
Comment (Quentas) [inline my_set/main/views.py:273]: do not put this info into repository

File: my_set/main/views.py
Comment (Quentas) [inline my_set/main/views.py:1]: views.py is used only for accepting request, running other functions and returning response. Functions about generating tokens, sending emails etc. should be moved to other file

File: my_set/main/views.py [hunk 13]
```diff
@@ -38,17 +43,234 @@ def loginPage(request):
...
+        decrypted_data = cryptocode.decrypt(token, 'django-insecure-svwex%*s7dnpav#)etq79gq4f+euje83rtwk9wduj=0f!l6m1-m')
+        data = json.loads(decrypted_data)
+        set_id = data['set_id']
+        set_instance = get_object_or_404(MySets, id=set_id)
+        set_instance.allowed_for.add(request.user)
+        return redirect('set', set_id=set_id)
+    except Exception as e:
+        return HttpResponseForbidden(e)
+
+def send_message(request, ):
```
Comment (Quentas) [inline my_set/main/views.py:260]: wtf is this?

File: my_set/main/views.py
Comment (Quentas) [inline my_set/main/views.py:1]: Add type annotations to all functions

**Review by Quentas — CHANGES_REQUESTED** [review]

File: my_set/main/models.py (hunk 8 above)
Comment (FFkmmr) [reply my_set/main/models.py:20]: I wanted to use it, but apparently I used a different method

**Pull request merged**